### PR TITLE
LGA 3239 Create Civil Cases model

### PR DESCRIPTION
**Pull request by BenMillar-MOJ**
### Adds the Civil Cases model
This model consists of:
- A UUID4 ID
- Created at and updated at timestamps
- A new required case_type field, which is an Enum
- A new optional assigned_to field, which is a VarChar

### Adds the BaseTableMixin class
This Mixin defines the fields which are shared across all database models, these include:
- ID, A UUID4 ID
- Created At, a timestamp which is set when the object is first instantiated
- Updated At, a timestamp which is set when the object is first instantiated and updated each time it is committed to the database.

This class also gives sub-models a default tablename of their model name, in lowercase, with an s.
Therefore the Case model will create the cases table, the EligibilityCheck model will create the eligibilitychecks table, etc.
This can be overwritten by subclasses.

### Adds the CaseType enum
This defines the allowed inputs for CaseType, currently this is just CLA and CCQ.
We use the full service names for the enum value to avoid confusion for potential consumers of the API that will be outside of the Info and Advice space.

## Any other changes that would benefit highlighting?

- Auto checks for UUID4 collisions when the database is being written to, will generate new UUID's if a collision would occur.

- Renames the Users model to User for consistency, this means we use the singular when dealing with individual objects and the plural when dealing with sets of those objects.

- Adds the `PersonalDataField`, this is subclass of `Field` which can be used to denote fields contain personal data. This will help with ensuring we manage personal data correctly as we introduce new features.

## Checklist

- [x] Provided JIRA ticket number in the title, e.g. "LGA-152: Sample title"

### CODE REVIEW

File: app/models/cases.py [hunk 1]
```diff
@@ -1,21 +1,22 @@
...
+from app.models.case_types import CaseTypes
 
 
 class BaseCase(SQLModel):
-    category: Categories = Field(index=True)
-    name: str
+    case_type: CaseTypes = Field(
+        index=True
+    )  # Which service is the case originally from
+    assigned_to: str | None = Field(
```
Comment (Pheonnexx) [inline app/models/cases.py:10]: we have removed 'assigned_to' to a seperate table - sorry!

**Review by Pheonnexx — CHANGES_REQUESTED** [review]

File: app/models/cases.py (hunk 1 above)
Comment (BenMillar-MOJ) [reply app/models/cases.py]: Thank you! I have removed assigned_to from this model with the latest commit!

File: app/models/users.py [hunk 2]
```diff
@@ -14,12 +14,14 @@ class TokenData(SQLModel):
     username: str | None = None
 
 
-class Users(SQLModel, table=True):
+class User(SQLModel, table=True):
```
Comment (said-moj) [inline app/models/users.py:17]: Should a user have an ID and when they were created?
Also it would be good to know when they last logged in so that we can monitor/delete in-active users

File: tests/cases/test_timestamp.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+def test_timezone():
+    case = Case(case_type=CaseTypes.CLA)
+    assert case.created_at.tzinfo == UTC
+
+
+def test_created_at():
+    before_case_creation = datetime.now(UTC)
+    case = Case(case_type=CaseTypes.CLA)
+    after_case_creation = datetime.now(UTC)
+    assert before_case_creation < case.created_at < after_case_creation
```
Comment (digitalali-moj) [inline tests/cases/test_timestamp.py:16]: Should we use <= here as well?

File: tests/cases/test_uuid.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    new_case = Case(id=original_id, case_type=CaseTypes.CLA)
+    session.add(new_case)
+    session.commit()
+    assert new_case.id != original_case.id  # Ensure the new case gets a new ID
+    assert (
+        original_case.id == original_id
+    )  # Ensure the original case keeps its original ID.
+
+
+def test_invalid_uuid_int(session: Session):
```
Comment (digitalali-moj) [inline tests/cases/test_uuid.py:50]: This is amazing.

I noticed you're testing validation, would it help to have something like that for the case type?

Also how about testing routes / responses? Or is that part of a separate scope?

**Review by digitalali-moj — COMMENTED** [review]
Amazing work thank you so much! Couple of questions only

File: tests/cases/test_uuid.py (hunk 4 above)
Comment (Pheonnexx) [reply tests/cases/test_uuid.py:50]: routes and responses are not part of this card - this i just implementing the model.

File: tests/cases/test_timestamp.py (hunk 3 above)
Comment (Pheonnexx) [reply tests/cases/test_timestamp.py:16]: no we shouldnt, as the time shouldnt be the same, so `before_case_creation` should be less than (and not equal too as they are not created at the same time `case.created_at` and then `case.created_at` should be less than `after_case_creation` as once again they are not created at the same and we are confirming that its both less and not equal to.

File: app/db/migrations/versions/2a0a7fdef0f6_add_users_to_db.py [hunk 5]
```diff
@@ -1,40 +1,34 @@
...
+branch_labels = None
+depends_on = None
 
 
 def upgrade() -> None:
-    # ### commands auto generated by Alembic - please adjust! ###
     op.create_table(
         "users",
         sa.Column("username", sqlmodel.sql.sqltypes.AutoString(), nullable=False),
         sa.Column("hashed_password", sqlmodel.sql.sqltypes.AutoString(), nullable=True),
```
Comment (TawneeOwl) [inline app/db/migrations/versions/2a0a7fdef0f6_add_users_to_db.py:25]: Can nullable be set to False please so a user can't be added with a null password.

File: app/models/users.py [hunk 6]
```diff
@@ -14,12 +14,14 @@ class TokenData(SQLModel):
...
     """
     Users are required to be authenticated to use some functionality of the API.
     Disabled users are unable to authenticate to receive a token.
     """
 
+    __tablename__ = "users"
+
     username: str = Field(primary_key=True)
     hashed_password: str | None = None
```
Comment (TawneeOwl) [inline app/models/users.py:26]: Can this be set to only being a str, no choice of being None as reference in the above; removing the null choice for password.

File: tests/cases/test_timestamp.py (hunk 3 above)
Comment (digitalali-moj) [reply tests/cases/test_timestamp.py:16]: How is this different from the test after where we use <= - just considering edge case if it's all within the same millisecond if that's even possible.

File: tests/cases/test_uuid.py (hunk 4 above)
Comment (digitalali-moj) [reply tests/cases/test_uuid.py:50]: Is there a way to feed in incorrect case type and should that be tested?

File: tests/cases/test_timestamp.py (hunk 3 above)
Comment (Pheonnexx) [reply tests/cases/test_timestamp.py:16]: discussed and resolve

**Review by said-moj — APPROVED** [review]
